helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -250,6 +253,23 @@ public void dropInstance(String clusterName, InstanceConfig instanceConfig) {
     }
   }
 
+  @Override
+  public List<String> purgeOfflineInstances(String clusterName, long offlineTimeout) {
+    Map<String, InstanceConfig> timeoutOfflineInstances = findTimeoutOfflineInstances(clusterName
+        , offlineTimeout);
+    List<String> failToPurgeInstances = new ArrayList<>();
+    timeoutOfflineInstances.values().forEach(instance -> {
+      try {
+        dropInstance(clusterName, instance);
+      } catch (HelixException e) {
+        logger.info("Failed to purge instance {} in cluster {}. Exception: {}", instance,
+            clusterName, e);
+        failToPurgeInstances.add(instance.getInstanceName());
+      }
+    });
+    return failToPurgeInstances;
```

Comment:
I think it's a bit awkward to return the list of instances whose deletion _failed_; usually these methods don't return anything, or they return the list of instances whose deletion _succeeded_.

how about we keep it as `void`, and only log one line at the end if there are any instances whose deletion failed? that way we persist the information we need while being concise.